magpie/services.py

```diff
@@ -412,21 +417,40 @@ class ServiceOWS(ServiceInterface):
 
     def __init__(self, service, request):
         # type: (models.Service, Request) -> None
-        super(ServiceOWS, self).__init__(service, request)
+        self._request = None
+        self.parser = None
+        super(ServiceOWS, self).__init__(service, request)  # sets request, which in turn parses it with below setter
+
+    def _get_request(self):
+        # type: () -> Request
+        return self._request
+
+    def _set_request(self, request):
+        # type: (Request) -> None
+        self._request = request
+        # must reset the parser from scratch if request changes to ensure everything is updated with new inputs
         self.parser = ows_parser_factory(request)
         self.parser.parse(self.params_expected)  # run parsing to obtain guaranteed lowered-name parameters
 
+    request = property(_get_request, _set_request)
```

Comment:
So, does this new property fix this issue https://github.com/Ouranosinc/Magpie/issues/433 ?
It looks like this is the only thing not related to improve exception reporting...

Reply:
I've relaunched the stress test on a daccs-iac instance with [bird-house/birdhouse-deploy@cache-settings]( https://github.com/bird-house/birdhouse-deploy/tree/cache-settings) and still see the 401. I cannot replicate though when only using `master` without cache. I'm not sure yet if they are related or not. Caching only seems to make it easier to catch the 401 since it will repeat over multiple requests rather than one-off call.

Reply:
With the new exception information, do we have a better clue on what is happening? What is the request params and why Magpie doesn't find the permission with it?

Reply:
I think I might have found the cause. 

When service is retrieved from the cache in MagpieAdapter, and gets forward to `effective_permissions` (method that resolves access with ACL), it tries to use that service. 
There is no database session associated to that service (since from cache), so somewhere deep in ziggurat_foundation code, `session.query` with `session=None` gets called, and raises `AttributeError`. 
Because of all the try/except, Twitcher simply catches it as returns "security check failed", can because the user happen to be anonymous, it indicates 401. 
What is tricky is that when the ACL itself was cached, this check was bypassed, which is why 200 was returned as normal. 
Same goes for the first request since service is not cached, so `effective_permisisons` are resolved with a valid session. 

Really a confusion time-specific situation!

Reply:
Should be adjusted with db-reconnect to service instance (see 0f99bdad0dc97762c97657c42f61b6f0b5e698d0)